apps/src/blockly/addons/cdoFieldAngleDropdown.ts

```diff
@@ -1,118 +1,177 @@
+      direction = sourceBlock.getFieldValue(this.directionFieldName);
     }
+    return direction || 'turnRight';
+  }
 
-    // Get a valid angle value, then find the nearest custom option.
-    const validAngleValue = super.doClassValidation_(newValue);
-
-    // A field wouldn't normally have both a menu generator and xml config values.
-    // If both are present, we favor the config values as they come directly
-    // from the serialization.
-    const customValues = this.configValues || this.menuGeneratorValues;
-    const closestCustomValue = customValues.reduce(
-      (prev: number, curr: number) => {
-        return Math.abs(curr - validAngleValue) <
-          Math.abs(prev - validAngleValue)
-          ? curr
-          : prev;
-      }
+  /**
+   * Create the container element for the angle helper.
+   * @returns {HTMLDivElement} The angle helper container element.
+   * @private
+   */
+  private createAngleHelperContainer(): HTMLDivElement {
+    const angleHelperContainer = document.createElement('div');
+    Blockly.utils.dom.addClass(
+      angleHelperContainer,
+      'blocklyAngleHelperContainer'
     );
-    return closestCustomValue;
+    return angleHelperContainer;
   }
 
   /**
-   * Converts xml element into dropdown field
-   * @param {Element} element xml field
-   * @override
+   * Initialize the angle helper.
+   * @private
    */
-  fromXml(element: Element) {
-    // If the field xml contains a `config`, then we can use it to limit the valid selectable options.
-    this.config = element.getAttribute('config');
-    if (this.config) {
-      this.configValues = this.config.split(',').map(Number);
-      this.min_ = Math.min(...this.configValues);
-      this.max_ = Math.max(...this.configValues);
-    }
-    super.fromXml(element);
+  private initializeAngleHelper(): void {
+    const container = this.createAngleHelperContainer();
+    const sourceBlock = this.getSourceBlock() as Block;
+    this.angleHelper = new Blockly.AngleHelper(this.getDirection(), {
+      onUpdate: this.updateDropdownMenuOptions.bind(this),
+      snapPoints: this.getOptions().map(option => parseInt(option[1])),
+      arcColour: (sourceBlock as ExtendedBlockSvg)?.style.colourPrimary,
+      angle: parseInt(`${this.getValue()}`),
+      enableBackgroundRotation: true,
+    });
+    Blockly.DropDownDiv.getContentDiv().appendChild(container);
+    this.angleHelper.init(container);
   }
 
   /**
-   * Store config options into xml element
-   * @param {Element} element xml field
-   * @return element
-   * @override
+   * Update the dropdown menu options to reflect the selected angle value.
+   * @param {string | number} angleValue The selected angle value.
+   * @private
    */
-  toXml(element: Element) {
-    if (this.config) {
-      element.setAttribute('config', this.config);
+  private updateDropdownMenuOptions(): void {
+    const angleValue = this.angleHelper?.getAngle() as number;
+    this.setValue(`${angleValue}`);
```

Comment:
nit: convert `angleValue` to a string when we initialize the value so we don't have to repeatedly convert it to a string in this method?